### PR TITLE
Add company backend for slack

### PR DESCRIPTION
**Pull request by zvbuhl**
This is a lightweight company-mode backend which adds user completion to slack. Adding channel completion should be relatively easy too.

It doesn't yet support @here.

### CODE REVIEW

**Comment by yuya373** [comment]
thanks! please add `slack-message-edit-mode` like this.

```
(defun company-slack-backend (command &optional arg &rest ignored)
       "Completion backend for slack chats.  It currently understands
@USER; adding #CHANNEL should be a simple matter of programming."
       (interactive (list 'interactive))
       (cl-case command
         (interactive (company-begin-backend 'company-slack-backend))
         (prefix (when (and (or (eq major-mode 'slack-mode)
                                (eq major-mode 'slack-edit-message-mode)) (looking-back "\\W@\\(\\w*\\)")) (match-string 1)))
         (candidates (remove-if-not
                      (lambda (x)
                        (s-starts-with-p arg x))
                      (mapcar #'first (slack-user-names slack-current-team))))
         (meta (format "%s%s"
                       (slack-user-presence-to-string (slack-user-find-by-name arg slack-current-team))
                       arg))))
```

**Comment by yuya373** [comment]
and It is better to leave adding `company-slack-backends` or not to user, I think.

**Comment by zvbuhl** [comment]
7ce2d4 adds completion to slack-edit-message mode.

I can remove the auto-adding of `company-slack-backend` to `company-backends`, but if the user's already loading `company` and `slack-mode` then he almost certainly wants completion, right?

**Comment by yuya373** [comment]
when adding completion-function to company-backends, there is some options like grouped backends.
users should be able to customize how to adding completion-function to company-backends.

File: slack-util.el [hunk 1]
```diff
@@ -55,5 +55,24 @@
...
+     (defun company-slack-backend (command &optional arg &rest ignored)
+       "Completion backend for slack chats.  It currently understands
+@USER; adding #CHANNEL should be a simple matter of programming."
+       (interactive (list 'interactive))
+       (cl-case command
+         (interactive (company-begin-backend 'company-slack-backend))
+         (prefix (when (and (eq major-mode 'slack-mode) (looking-back "\\W@\\(\\w*\\)")) (match-string 1)))
+         (candidates (remove-if-not
+                      (lambda (x)
+                        (s-starts-with-p arg x))
```
Comment (yuya373) [inline slack-util.el]: there is no dependency to `s.el` in this package. please replace `s-starts-with-p` with `string-prefix-p`.

**Comment by zvbuhl** [comment]
I made company-mode optional and removed that dependency.

**Comment by yuya373** [comment]
thanks!  very appreciate for contributing!

**Pull request merged**